kernel/src/hil/display.rs

```diff
@@ -0,0 +1,376 @@
+    fn print(
+        &self,
+        buffer: &'static mut [u8],
+        len: usize,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Sends to the driver a command to set the cursor at a given position
+    /// (x_position, y_position). When finished, the driver will call the
+    /// `command_complete()` callback.
+    ///
+    /// Return values:
+    /// - `Ok(())`: The command is valid and will be sent to the driver.
+    /// - `BUSY`: Another command is in progress.
+    fn set_cursor(&self, x_position: usize, y_position: usize) -> Result<(), ErrorCode>;
+
+    /// Sends to the driver a command to show the cursor. When finished,
+    /// the driver will call the `command_complete()` callback.
+    ///
+    /// Return values:
+    /// - `Ok(())`: The command is valid and will be sent to the driver.
+    /// - `BUSY`: Another command is in progress.
+    fn set_show_cursor(&self, show: bool) -> Result<(), ErrorCode>;
+
+    /// Sends to the driver a command to turn on the blinking cursor. When finished,
+    /// the driver will call the `command_complete()` callback.
+    ///
+    /// Return values:
+    /// - `Ok(())`: The command is valid and will be sent to the driver.
+    /// - `BUSY`: Another command is in progress.
+    fn set_blink_cursor(&self, blink: bool) -> Result<(), ErrorCode>;
+
+    /// Sends to the driver a command to clear the display of the screen.
+    /// When finished, the driver will call the `command_complete()` callback.
+    ///
+    /// Return values:
+    /// - `Ok(())`: The command is valid and will be sent to the driver.
+    /// - `BUSY`: Another command is in progress.
+    fn clear(&self) -> Result<(), ErrorCode>;
+}
+
+pub trait TextBufferClient {
+    /// The driver calls this function when any command (but a write one)
+    /// finishes executing.
+    fn command_complete(&self, r: Result<(), ErrorCode>);
+
+    /// The driver calls this function when a write command finishes executing.
+    fn write_complete(&self, buffer: &'static mut [u8], len: usize, r: Result<(), ErrorCode>);
+}
+
+/* FrameBuffer */
+
+pub trait FrameBuffer<'a> {
+    /// Returns a tuple (width, height) with the current resolution (in pixels)
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    ///
+    /// note that width and height may change due to rotation
+    fn get_mode(&self) -> GraphicsMode;
+
+    /// Returns the minim size (in bytes) that can be written to the framebuffer.
+    ///
+    /// Due to hardware constraints, framebuffer writes have to be rounded up to a tile size.
+    /// This means that the size of the write buffer has to be a multiple of the tile. An
+    /// example use case is a framebuffer that has the minimum write unit a full line on
+    /// the display. This means that clients can only write entier lines to the framebuffer
+    /// as opossed to single pixels.
+    fn get_block_size(&self) -> usize;
```

Comment:
There are at least 4 coordinates needed for tiling: xoffset, yoffset, xsize, ysize. The tiles also need to get internal pixel ordering: some display drivers (PCD8544) will iterate rows within a tile first, and then columns.

Reply:
Thought through in https://github.com/tock/tock/pull/3067